toolchain/semantics/semantics_context.h

```diff
@@ -161,6 +171,38 @@ class SemanticsContext {
     Compatible,
   };
 
+  // Provides DenseMap support for canonical_struct_types_.
+  struct StructTypeMapInfo {
+    using KeyType =
+        llvm::SmallVector<std::pair<SemanticsStringId, SemanticsTypeId>>;
+    using IntMapInfo = llvm::DenseMapInfo<int32_t>;
+
+    static inline auto getEmptyKey() -> KeyType {
+      return {{SemanticsStringId(IntMapInfo::getEmptyKey()),
+               SemanticsTypeId(IntMapInfo::getEmptyKey())}};
+    }
+
+    static inline auto getTombstoneKey() -> KeyType {
+      return {{SemanticsStringId(IntMapInfo::getTombstoneKey()),
+               SemanticsTypeId(IntMapInfo::getTombstoneKey())}};
+    }
+
+    static auto getHashValue(const KeyType& val) -> unsigned {
+      unsigned hash_value = 0;
+      for (const auto& [string_id, type_id] : val) {
+        hash_value = llvm::detail::combineHashValue(
+            hash_value, IntMapInfo::getHashValue(string_id.index));
+        hash_value = llvm::detail::combineHashValue(
+            hash_value, IntMapInfo::getHashValue(type_id.index));
+      }
+      return hash_value;
```

Comment:
If you do keep this approach, you should be able to use the range API to get something much more efficient w/o much effort: https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/Hashing.h#L484-L493

Reply:
Switched to FoldingSet.